consul/consul.go

```diff
@@ -137,15 +134,15 @@ func (c *Consul) Register(service *registry.Service) {
 //
 func (c *Consul) Deregister() error {
 	for s, b := range serviceCache {
-		if !b.isRegistered {
+		if c.CacheIsValid(s) {
+			c.CacheProcessDeregister(s)
+		} else {
```

Comment:
Right, so the first time the task is discovered it is registered and the cache starts at 0 < 1 in the start of the `Refresh()` cycle, the final part of the `Refresh()` cycle it iterates through all of the task and invalidates the cache (1 !< 1 now). So now the next `Refresh()` if the task is still present it marks the cache as valid (back to 0 < 1), as the last part of the `Refresh()` cycle it will once again invalidate the cache (1 !< 1 again). So let's say this time the task is gone, the `Refresh()` cycle will never mark that tasks cache as still valid during the registration portion (still stays at 1 !< 1), so now when `Refresh()` finally calls `Deregister()` it will remove the task from Consul.

Does that make sense? I can try to draw up a better example if it doesn't.